pkg/controllers/switchedcontroller/switched_controller.go

```diff
@@ -0,0 +1,127 @@
+	"github.com/openshift/library-go/pkg/operator/v1helpers"
+)
+
+type ControllerWithSwitch struct {
+	delegateName       string
+	delegateFactoryFn  DelegateFactoryFunc
+	delegateController factory.Controller
+
+	switchConditionFn   DelegateSwitchCondition
+	switchContext       context.Context
+	switchContextCancel context.CancelFunc
+
+	mutex sync.Mutex
+}
+
+// DelegateSwitchCondition defines a condition function that controls when the delegate
+// controller must be switched on or off.
+type DelegateSwitchCondition func() (bool, error)
+
+// DelegateFactoryFunc returns a controller factory that can be used to create an instance
+// of the delegate controller. The SwitchedController's context is passed to the function
+// so that it may be used to start any informers that also depend on the DelegateSwitchCondition.
+// This context is cancelled when DelegateSwitchCondition returns (false, nil).
+type DelegateFactoryFunc func(context.Context) *factory.Factory
+
+// NewControllerWithSwitch creates an instance of a switched controller. The switched controller is
+// defined by the following:
+// - delegateFactoryFn: a function that is invoked when a new instance of the delegate controller must be created
+// - switchConditionFn: a function that is invoked on every switched controller sync to determine whether it needs to switch on/off the delegate controller
+// - informers: any informers that must be tracked and are required by the switch condition
+//
+// If no delegateFactoryFn is defined, the controller's sync will always fail. If no switchConditionFn is defined,
+// the delegate controller will never be switched off.
+func NewControllerWithSwitch(
+	operatorClient v1helpers.OperatorClient,
+	delegateName string,
+	delegateFactoryFn DelegateFactoryFunc,
+	switchConditionFn DelegateSwitchCondition,
+	informers []factory.Informer,
+	resyncInterval time.Duration,
+	eventRecorder events.Recorder,
+) factory.Controller {
+
+	c := &ControllerWithSwitch{
+		delegateName:      delegateName,
+		delegateFactoryFn: delegateFactoryFn,
+		switchConditionFn: switchConditionFn,
+	}
+
+	return factory.New().
+		WithSync(c.sync).
+		WithSyncDegradedOnError(operatorClient).
+		WithInformers(informers...).
+		ResyncEvery(resyncInterval).
+		ToController(delegateName+"_SwitchedController", eventRecorder)
+}
+
+func (c *ControllerWithSwitch) runDelegate(ctx context.Context, syncCtx factory.SyncContext) {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	if c.delegateController == nil {
+		c.delegateController = c.delegateFactoryFn(ctx).ToController(c.delegateName, syncCtx.Recorder())
+	}
+	go c.delegateController.Run(ctx, 1)
+}
+
+func (c *ControllerWithSwitch) stopDelegate() {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	if c.switchContextCancel != nil {
+		c.switchContextCancel()
+	}
+
+	c.switchContext = nil
+	c.delegateController = nil
+}
+
+func (c *ControllerWithSwitch) sync(ctx context.Context, syncCtx factory.SyncContext) error {
```

Comment:
Is it safe for us to assume that the `ctx` parameter that is passed in here is a long lived context that isn't cancelled after the sync method completes?

If it does get cancelled after the sync method completes, it likely is not safe for us to use this context as the basis for starting and stopping the delegate controller as we would start it, return from `sync` and then it would be immediately stopped.